### PR TITLE
Validate Annotations present in image manifest and fallback to annota…

### LINKED ISSUE #788: Zot annotation validation - check labels are on the config, and fallback to annotations on the manifest if not available
Currently we only validate Annotations present in image manifest. Most images contain the same metadata in config labels.
We need to check both.

Most likely the manifest annotations should be checked first.

### PR DESCRIPTION
**Pull request by nicoldr**
…tions in the config if not available

Signed-off-by: Nicol Draghici <idraghic@cisco.com>

**What type of PR is this?**

**Which issue does this PR fix**:
Closes #788 

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/extensions/lint/lint.go [hunk 1]
```diff
@@ -53,17 +53,57 @@ func (linter *Linter) CheckMandatoryAnnotations(repo string, manifestDigest godi
...
-			// if annotations are not found, return false but it's not an error
-			linter.log.Error().Msgf("linter: missing %s annotations", annot)
+	configAnnotations := manifest.Config.Annotations
+	for annotation := range configAnnotations {
+		if _, ok := mandatoryAnnotationsMap[annotation]; ok {
+			mandatoryAnnotationsMap[annotation] = true
+		}
+	}
+
+	if !checkAllAnnotationsFound(mandatoryAnnotationsMap) {
```
Comment (peusebiu) [inline pkg/extensions/lint/lint.go:95]: What if we have a single function getMissingAnnotations(annotations, mandatoryAnnotations map[string]bool)
```
		for annotation, flag := range mandatoryAnnotationsMap {
			if !flag {
				missingAnnotations = append(missingAnnotations, annotation)
			}
		}
```

and the check after manifest would be if len(missingAnnotations) == 0 { return true, nil}
check after config would be 
```
              if len(missingAnnotations) > 0 {
                   		linter.log.Error().Msgf("linter: missing %s annotations:", missingAnnotations)
                                return false, nil
               } 
```

Comment (nicoldr) [reply pkg/extensions/lint/lint.go]: I implemented a new method as you suggested, replacing those checks in one method getMissingAnnotations()

File: pkg/extensions/lint/lint.go [hunk 2]
```diff
@@ -53,19 +53,57 @@ func (linter *Linter) CheckMandatoryAnnotations(repo string, manifestDigest godi
...
-			return false, nil
+	configAnnotations := manifest.Config.Annotations
+	for annotation := range configAnnotations {
+		if _, ok := mandatoryAnnotationsMap[annotation]; ok {
+			mandatoryAnnotationsMap[annotation] = true
 		}
 	}
 
+	missingAnnotations = getMissingAnnotations(mandatoryAnnotationsMap)
+	if len(missingAnnotations) == 0 {
```
Comment (peusebiu) [inline pkg/extensions/lint/lint.go:97]: remove this, no need for it

File: pkg/extensions/lint/lint.go [hunk 3]
```diff
@@ -53,19 +53,53 @@ func (linter *Linter) CheckMandatoryAnnotations(repo string, manifestDigest godi
...
-		if !found {
-			// if annotations are not found, return false but it's not an error
-			linter.log.Error().Msgf("linter: missing %s annotations", annot)
+	if err != nil {
+		linter.log.Error().Err(err).Msg("linter: couldn't get config JSON" + string(configDigest))
+
+		return false, err
+	}
+
+	if err := json.Unmarshal(content, &manifest); err != nil {
```
Comment (peusebiu) [inline pkg/extensions/lint/lint.go:83]: this should be config struct not manifest

File: pkg/extensions/lint/lint.go [hunk 4]
```diff
@@ -53,19 +53,53 @@ func (linter *Linter) CheckMandatoryAnnotations(repo string, manifestDigest godi
...
+		linter.log.Error().Err(err).Msg("linter: couldn't unmarshal config JSON")
 
-			return false, nil
+		return false, err
+	}
+
+	configAnnotations := manifest.Config.Annotations
```
Comment (peusebiu) [inline pkg/extensions/lint/lint.go:89]: here you are looking on config descriptor annotations, which is found in the manifest, we need to look on config labels from config struct (not its descriptor)

look at my pr: https://github.com/project-zot/zot/pull/787

so you need to unmarshal the config blob in a ispec.ImageConfig var and look on its labels.

Right @andaaron ?

File: pkg/extensions/lint/lint_test.go [hunk 5]
```diff
@@ -241,6 +242,60 @@ func TestVerifyMandatoryAnnotations(t *testing.T) {
...
+
+		manifestBlob := resp.Body()
+		var manifest ispec.Manifest
+		err = json.Unmarshal(manifestBlob, &manifest)
+		So(err, ShouldBeNil)
+
+		manifest.Annotations = make(map[string]string)
+
+		manifest.Annotations["annotation1"] = "testFail1"
+		manifest.Config.Annotations = make(map[string]string)
```
Comment (andaaron) [inline pkg/extensions/lint/lint_test.go:286]: Needs to be updated according to Sebi's comment, to verify an object of type  ispec.ImageConfig

**Review by peusebiu — APPROVED** [review]
LGTM

**Review by andaaron — APPROVED** [review]
Looks good.

**Pull request merged**